hash_practice/exercises.py

```diff
@@ -2,18 +2,65 @@
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n)
+        Space Complexity: O(n)
     """
-    pass
+    if len(strings) == 0:
+        return []
+
+    hash = {}
+    for string in strings:
+        #sort out the word alphabetically since we are working with anagrams and make that the key
+        #sorted returns a list and we need to change it back to a string by using join
+        key = ''.join(sorted(string))
+        if key not in hash:
+            #bracket notation to add it in as a list
+            #we're adding in the actual, not sorted, version of the string
+            hash[key] = [string]
+        else:
+            #since the list is already there, append to it
+            hash[key].append(string)
+    print(hash.values())
+    return hash.values()
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n)
+        Space Complexity: O(n)
     """
-    pass
+    #edge case
+    if len(nums) == 0:
+        return []
+
+    #create a hashmap
+    hash = {}
+
+    #Build the hashmap for the elements
+    for value in nums:
+        #if the value is in the hashmap, increment it

```

Comment:
Likewise, while comments like these saying _what_ the code does can be very useful when first writing, it's usually a good idea to get rid of them once the code is done. This is because your code is the real implementation of _what_ the code does, and the comment may eventually fall out of sync with the implementation and lead to confusion when your code is being read.

The way that I think of it is that instead only keep _why_ comments that describes _why_ the code does what it does. You can generally trust that other readers of your code will understand _what_ the code does from the code, but understanding _why_ the code does that is a much more valuable use of a comment.

Reply:
(Obviously documentation comments (`"""`) can be more about _what_ the code does, but that's an exception.)